nativelib/src/main/resources/process_monitor.c

```diff
@@ -0,0 +1,102 @@
+            return res;                                                        \
+    } while (0)
+
+struct proc_info {
+    int pid;
+    int result;
+};
+static pthread_mutex_t shared_mutex;
+static pthread_mutex_t private_mutex;
+static pthread_cond_t cond;
+static pthread_t main_thread;
+static int last_pid = -1;
+static int last_result = -1;
+
+void *wait_loop(void *arg) {
+    int status;
+    // Block all signals on the monitor thread.
+    sigset_t set;
+    sigfillset(&set);
+    pthread_sigmask(SIG_BLOCK, &set, NULL);
+
+    pthread_mutex_lock(&private_mutex);
+    while (1) {
+        int pid = waitpid(-1, &status, 0);
+        if (pid != -1) {
+            pthread_mutex_lock(&shared_mutex);
+            last_pid = pid;
+            last_result = WIFSIGNALED(status) ? 0x80 + status : status;
+            pthread_kill(main_thread, SIGUSR1);
+            // Wait until the main thread has handled the signal to start
+            // polling for pids again.
+            pthread_cond_wait(&cond, &private_mutex);
+            pthread_mutex_unlock(&shared_mutex);
+        }
+    }
+    // should be unreachable
+    pthread_mutex_unlock(&private_mutex);
+    return NULL;
+}
+
+void scalanative_process_monitor_shared_mutex(pthread_mutex_t **mutex) {
+    *mutex = &shared_mutex;
+}
+
+int scalanative_process_monitor_signal_cond(pthread_cond_t *proc_cond) {
+    return pthread_cond_broadcast(proc_cond);
+}
+
+int scalanative_process_monitor_mutex_lock(pthread_mutex_t *mutex) {
+    return pthread_mutex_lock(mutex);
+}
+
+int scalanative_process_monitor_mutex_unlock(pthread_mutex_t *mutex) {
+    return pthread_mutex_unlock(mutex);
+}
+
+void scalanative_process_monitor_last_proc_info(struct proc_info *info) {
+    info->pid = last_pid;
+    info->result = last_result;
+}
+
+int scalanative_process_monitor_notify() {
+    RETURN_ON_ERROR(pthread_mutex_lock(&private_mutex));
+    RETURN_ON_ERROR(pthread_cond_broadcast(&cond));
+    return pthread_mutex_unlock(&private_mutex);
+}
+
+int scalanative_process_monitor_wait(pthread_mutex_t *mutex,
+                                     pthread_cond_t **cond_p,
+                                     struct timespec *ts) {
+    int res = 0;
+    pthread_cond_t cond;
+
+    RETURN_ON_ERROR(pthread_cond_init(&cond, NULL));
+    *cond_p = &cond;
+    // The expectation is that this call will exit due to an interrupt rather
+    // than a signal/broadcast.
+    res = ts ? pthread_cond_timedwait(&cond, mutex, ts)
+             : pthread_cond_wait(&cond, &shared_mutex);
+    pthread_cond_destroy(&cond);
+    return res;
+}
+
+void scalanative_process_monitor_init() {
+    pthread_t thread;
+    pthread_mutex_init(&shared_mutex, NULL);
+    pthread_mutex_init(&private_mutex, NULL);
+    pthread_cond_init(&cond, NULL);
+    pthread_create(&thread, NULL, wait_loop, NULL);
+    pthread_detach(thread);
+    main_thread = pthread_self();
+}
```

Comment:
This file needs to be split into in 2: forwarder methods for posix, and another one for "runtime" code for process implementation.